hooks/tk-multi-publish2/maya/collector_xgen_dependencies.py

```diff
@@ -0,0 +1,348 @@
+
+                # Adding xgen data path in collector
+                xgPatchCache = palette_info[palette]['xgPatchCache']
+                if os.path.isfile(xgPatchCache):
+                    dict_values.update(self.resolve_custom_path_template(
+                        settings, parent_item, 'file.xgen.patch', xgPatchCache))
+                    self.logger.info(" Collected file.xgen.patch: '%s'" % (os.path.basename(xgPatchCache)))
+                else:
+                    error_msg = "Collect Xgen Dependencies failed. " \
+                                "Missing xgen patch caches: '{}'".format(xgPatchCache)
+                    self.logger.error(error_msg)
+                    raise Exception(error_msg)
+
+                xgen_dependencies[str(palette)] = dict_values
+            parent_item.properties.xgen_dependencies = xgen_dependencies
+        return None
+
+    def resolve_custom_path_template(self, settings, parent_item, item_key, work_path):
+        """
+        Resolve custom work and publish path_template from the given work_path for the
+        specified item.
+
+        :param settings: collector settings
+        :param parent_item: parent item
+        :param item_key: valid settings item key
+        :param work_path: source work path to resolve publish type and publish path
+        :return: Resolved values in dict format: {publish_type:{work_path:, publish_path:}}
+        """
+
+        resolved_values = dict()
+
+        work_tmpl = self.sgtk.template_from_path(work_path)
+        fields = work_tmpl.get_fields(work_path)
+
+        if item_key == 'file.xgen.collection':
+            """
+            By default xgen collection doesn't maintain versions 
+            and link with source maya file. So to resolve this issue, 
+            We are collecting session item fields and passing the 
+            missing values to xgen collections publish template 
+            fields which creates the folder with session file name 
+            and publishes collections inside.
+            """
+            # Resolve session work path template
+            session_work_path = parent_item.properties.path
+            session_work_template = self.sgtk.template_from_path(session_work_path)
+            session_item_fields = session_work_template.get_fields(session_work_path)
+
+            # Adding required session name fields
+            fields['Step'] = session_item_fields['Step']
+            fields['name'] = session_item_fields['name']
+            fields['version'] = session_item_fields['version']
+
+        publish_tmp_name = settings[item_key]['publish_path_template'].value
+        publish_tmp_obj = self.sgtk.templates[publish_tmp_name]
+        publish_path = publish_tmp_obj.apply_fields(fields)
+        publish_type = settings[item_key]['publish_type'].value
+
+        resolved_values[publish_type] = {'work_path': str(work_path), 'publish_path': str(publish_path)}
+
+        return resolved_values
```

Comment:
Let's restrict figuring out `publish_type` and `publish_path` to the publish plugin instead of resolving those values in the collector. The above properties should be figured out in validate, please look at `color_process_files` hook as an example.
Collector should ideally care more about whether a `work_path_template` matches the paths that you have collected or not.